### PR TITLE
Upgrade procedure

### PR DESCRIPTION
**Pull request by stephenrlouie**
- Cluster and gem master procedure
- Does not clean up old versions on the gem master
- Entire destruction of the cluster is needed.

Pending on this PR with contrib: https://github.com/kubernetes/contrib/pull/870

### CODE REVIEW

File: docs/UPGRADES.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#Update / Downgrade Procedure
```
Comment (danehans) [inline docs/UPGRADES.md]: s/Update/Upgrade

File: docs/UPGRADES.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+#Uninstall the current cluster configuration:
+```
+cd contrib/ansible
+ansible-playbook -i inventory playbooks/adhoc/uninstall.yml
+```
```
Comment (danehans) [inline docs/UPGRADES.md]: I don't believe this is how we should upgrade the cluster. Uninstall essentially wipes out the cluster and then the cluster is being rebuilt from scratch using the new bin version of the service. We should take a fine-grained approach to upgrades. At a high level, this is needed:

1. Create an upgrade playbook for each service that we will support upgrades for, starting with etcd: contrib/ansible/playbooks/adhoc/upgrade_etcd.yml
2. You can use playbooks/adhoc/uninstall.yml as a reference for creating playbooks/adhoc/upgrade.yml
3. upgrade_etcd.yml should contain a subset of tasks from contrib/ansible/roles/etc/tasks and run against the [etcd] host group.

Spend time reviewing all the tasks in contrib/ansible/roles/etc/tasks . Copy the code that is related to downloading the etcd bin, creating the bin dir structure and restarting the service.

Then rinse and repeat for other services we will provide upgrade support for. Lastly, create upgrade.sh that runs the appropriate upgrade playbook(s)

When this is complete, extend the upgrade functionality to gemini-project/gemini

Comment (stephenrlouie) [inline docs/UPGRADES.md]: With contrib as it is right now, would it be better just to run `setup.sh` again... I think that should work if you update version numbers, testing now.

I will also be looking at contrib to make an upgrade playbook in `adhoc`

Comment (stephenrlouie) [inline docs/UPGRADES.md]: Running setup.sh again as a downgrade procedure is not advisable. 

Kube 1.2.2 -> 1.2.0 (Only downgrades the client)
Flannel 0.5.5 -> 0.5.4 (Downgrades)
Pypy 2.4.0 -> 2.3.1(Did not downgrade)

Init:
```
KUBE: 
Client Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.2", GitCommit:"528f879e7d3790ea4287687ef0ab3f2a01cc2718", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.2", GitCommit:"528f879e7d3790ea4287687ef0ab3f2a01cc2718", GitTreeState:"clean"}
FLANNEL: 
0.5.5
PYTHON: 
/opt/bin/pypy/bin/pypy: /lib64/libssl.so.1.0.0: no version information available (required by /opt/bin/pypy/bin/pypy)
/opt/bin/pypy/bin/pypy: /lib64/libcrypto.so.1.0.0: no version information available (required by /opt/bin/pypy/bin/pypy)
Python 2.7.8 (f5dcc2477b97, Sep 18 2014, 11:33:30)
[PyPy 2.4.0 with GCC 4.6.3]
```

Downgrade:
```
KUBE: 
Client Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.0", GitCommit:"5cb86ee022267586db386f62781338b0483733b3", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.2", GitCommit:"528f879e7d3790ea4287687ef0ab3f2a01cc2718", GitTreeState:"clean"}
FLANNEL: 
0.5.4
PYTHON: 
/opt/bin/pypy/bin/pypy: /lib64/libssl.so.1.0.0: no version information available (required by /opt/bin/pypy/bin/pypy)
/opt/bin/pypy/bin/pypy: /lib64/libcrypto.so.1.0.0: no version information available (required by /opt/bin/pypy/bin/pypy)
Python 2.7.8 (f5dcc2477b97, Sep 18 2014, 11:33:30)
[PyPy 2.4.0 with GCC 4.6.3]
```


Let's leave this open, I'll look into `contrib/ansible/adhoc` and add upgrade / downgrade playbooks and we can rewrite the procedure once thats done

**Comment by stephenrlouie** [comment]
Need to refactor update the upstream stuff first, like this [etcd refactor] (https://github.com/kubernetes/contrib/pull/982). Once that happens we can re-open and integrate the upgrade story

**Pull request closed without merging**